design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,597 @@
+        }
+        rust(".build()")
+    }
+}
+```
+
+</details>
+
+#### Loading values for builtIns
+
+The fundamental point of builtIn values is enabling _other codegenerators to define where these values come from.
+Because of that,
+we will need to expose the ability to customize builtIns. One way to do this is with a new customization
+type, `EndpointCustomization`:
+
+```kotlin
+fun endpointCustomizations(
+    clientCodegenContext: C,
+    operation: OperationShape,
+    baseCustomizations: List<EndpointCustomization>
+): List<EndpointCustomization> = baseCustomizations
+
+
+abstract class EndpointCustomization {
+    abstract fun defaultFor(parameter: Parameter, config: String): Writable?
+}
+```
+
+Customizations have the ability to specify the default value for a parameter. (Of course, these customizations need to
+be wired in properly.)
+
+### Converting a Smithy Endpoint to an AWS Endpoint
+
+A Smithy endpoint has an untyped, string->`Document` collection of properties. We need to interpret these properties to
+handle actually resolving an endpoint. We will implement this by implementing `TryFrom` from the Smithy endpoint to the
+AWS endpoint:
+> Note: It's possible `AWSEndpoint` will be removed entirely and the concept will be solely expressed in middleware. If
+> this is the case, this code would be ported into middleware directly.
+
+<details>
+<summary>Converting from smithy-native `Endpoint` to `AwsEndpoint`</summary>
+
+```rust
+impl TryFrom<Endpoint> for AwsEndpoint {
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+
+    fn try_from(value: Endpoint) -> Result<Self, Self::Error> {
+        let uri: Uri = value.url().parse()?;
+        let base_endpoint = SmithyEndpoint::mutable(uri);
+        // look for v4 as an auth scheme
+        let auth_schemes = match value
+            .properties()
+            .get("authSchemes")
+            .ok_or("no auth schemes in metadata")?
+        {
+            Document::Array(schemes) => schemes,
+            _other => return Err("expected an array for authSchemes".into()),
+        };
+        let v4 = auth_schemes
+            .iter()
+            .flat_map(|doc| match doc {
+                Document::Object(map)
+                if map.get("name") == Some(&Document::String("v4".to_string())) =>
+                    {
+                        Some(map)
+                    }
+                _ => None,
+            })
+            .next()
+            .ok_or("could not find v4 as an acceptable auth scheme")?;
+
+        let signing_scope = match v4
+            .get("signingScope")
+            .ok_or_else(|| format!("signing scope missing: {:?}", v4))?
+        {
+            Document::String(s) => s.clone(),
+            _ => return Err("expected error".into()),
+        };
+        let signing_service = match v4.get("signingName").ok_or("signing service missing")? {
+            Document::String(s) => s.clone(),
+            _ => return Err("expected error".into()),
```

Comment:
shouldn't these just use `map`? Can we actually hit these error cases?

Reply:
yeah this would happen if you used an invalid endpoint resolver